controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,357 @@
+	}
+	if config.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	switch config.Status.Phase {
+	case gkeConfigImportingPhase:
+		return h.importCluster(config)
+	case gkeConfigNotCreatedPhase:
+		return h.create(config)
+	case gkeConfigCreatingPhase:
+		return h.waitForCreationComplete(config)
+	case gkeConfigActivePhase:
+		return h.checkAndUpdate(config)
+	case gkeConfigUpdatingPhase:
+		return h.checkAndUpdate(config)
+	}
+
+	return config, nil
+}
+
+// recordError writes the error return by onChange to the failureMessage field on status. If there is no error, then
+// empty string will be written to status
+func (h *Handler) recordError(onChange func(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error)) func(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	return func(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+		var err error
+		var message string
+		config, err = onChange(key, config)
+		if config == nil {
+			// GKE config is likely deleting
+			return config, err
+		}
+		if config.Status.FailureMessage == message {
+			return config, err
+		}
+
+		config = config.DeepCopy()
+
+		if message != "" {
+			if config.Status.Phase == gkeConfigActivePhase {
+				// can assume an update is failing
+				config.Status.Phase = gkeConfigUpdatingPhase
+			}
+		}
+		config.Status.FailureMessage = message
+
+		var recordErr error
+		config, recordErr = h.gkeCC.UpdateStatus(config)
+		if recordErr != nil {
+			logrus.Errorf("Error recording gkecc [%s] failure message: %s", config.Name, recordErr.Error())
+		}
+		return config, err
+	}
+}
+
+// importCluster cluster returns a spec representing the upstream state of the cluster matching to the
+// given config's displayName and region.
+func (h *Handler) importCluster(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	config.Status.Phase = gkeConfigActivePhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) OnGkeConfigRemoved(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	if config.Spec.Imported {
+		logrus.Infof("cluster [%s] is imported, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+	if config.Status.Phase == gkeConfigNotCreatedPhase {
+		// The most likely context here is that the cluster already existed in GKE, so we shouldn't delete it
+		logrus.Warnf("cluster [%s] never advanced to creating status, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+
+	logrus.Infof("deleting cluster [%s]", config.Name)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := utils.GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	logrus.Debugf("Removing cluster %v from project %v, region/zone %v", config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region)
+	operation, err := utils.WaitClusterRemoveExp(ctx, svc, config)
```

Comment:
I think this is taken from kev1, but I prefer if we just use the ExponentialBackoff function from k8s `wait` package instead of rewriting it.

Reply:
I responded to this on the old PR https://github.com/rancher/gke-operator/pull/1#discussion_r580661841

I do want to clean this up but I'm not sure if exponential backoff is the best way to go.

Reply:
Read the comment, using the re-queueing pattern would be good if it works. As far as the exponential backoff function goes, if you set the factor to one it's really just a regular backoff. Also, the requeueing in the wrangler framework is an exponential backoff itself afaik.